### PR TITLE
[clang-tidy] Add `bugprone-pointer-arithmetic-on-polymorphic-object` check

### PR DESCRIPTION
**Pull request by Discookie**
Finds pointer arithmetic on classes that declare a virtual function.

This check corresponds to the SEI Cert rule [CTR56-CPP: Do not use pointer arithmetic on polymorphic objects](https://wiki.sei.cmu.edu/confluence/display/cplusplus/CTR56-CPP.+Do+not+use+pointer+arithmetic+on+polymorphic+objects).

```cpp
struct Base {
  virtual void ~Base();
};

struct Derived : public Base {};

void foo(Base *b) {
  b += 1; // passing `Derived` to `foo()` results in UB
}
```

[Results on open-source projects](https://codechecker-demo.eastus.cloudapp.azure.com/Default/runs?run=Discookie-ctr56-with-classnames). Most of the Qtbase reports are from having a `virtual override` declaration, and the LLVM reports are true positives, as far as I can tell.

### CODE REVIEW

**Review by EugeneZelenko — COMMENTED** [review]
Please mention new check in Release Notes.

**Review by 5chmidti — CHANGES_REQUESTED** [review]
Needs some small changes before landing. It's a good check to have, thanks.

File: clang-tools-extra/test/clang-tidy/checkers/bugprone/virtual-arithmetic.cpp
Comment (PiotrZSL) [inline clang-tools-extra/test/clang-tidy/checkers/bugprone/virtual-arithmetic.cpp:20]: add tests for ++b;

File: clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.cpp:44]: `on polymorphic class` ... also "if the pointee is a different class" is not so clear.
 Problem is that this issue may happen regardless if this is an polymorphic  object or not...

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/virtual-arithmetic.rst
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/bugprone/virtual-arithmetic.rst:10]: this apply not only to polymorphic objects. Would be good to write why (maybe something about pointer to virtual table).

File: clang-tools-extra/test/clang-tidy/checkers/bugprone/virtual-arithmetic.cpp
Comment (PiotrZSL) [inline clang-tools-extra/test/clang-tidy/checkers/bugprone/virtual-arithmetic.cpp:59]: Add test for bugfree scenario:
```
void test()
{
    Object table[10];
    Object* ptr = table[5];
}
```

**Review by PiotrZSL — COMMENTED** [review]
Few nits.
My main concern is an check name, had to open PR to find out what it does.
maybe should be more a "bugprone-arithmetic-on-polymorphic-pointer" or something like that.

**Comment by carlosgalvezp** [comment]
I'm not very happy about the name, there's no such thing as "virtual arithmetic".

What about "bugprone-pointer-arithmetic-polymorphic-object"?

I rather have a long descriptive name than a short ambiguous name.

**Comment by carlosgalvezp** [comment]
Otherwise since this is a problem also on non-polymorphic objects, maybe "bugprone-pointer-arithmetic-base-class"?

**Comment by PiotrZSL** [comment]
bugprone-pointer-arithmetic-polymorphic-object or bugprone-pointer-arithmetic-on-polymorphic-object could be fine.
limiting check to polymorphic objects it's fine, someone can always enable an cppguidelines check to cover all cases.

Also check-alias should be added to cert.

**Comment by carlosgalvezp** [comment]
I realize it's probably very hard (if not impossible) to detect whether a class is a base class unless it has virtual functions. 

That's why the corresponding AUTOSAR rule bans only non-final classes. But that is a bit overkill and leads to some pain.

File: clang-tools-extra/test/clang-tidy/checkers/bugprone/virtual-arithmetic.cpp
Comment (Discookie) [reply clang-tools-extra/test/clang-tidy/checkers/bugprone/virtual-arithmetic.cpp]: Is this snippet meant to be `Object &ref`? Array subscript can't return a pointer AFAIK.
Other than that the case is not too distinct from `Derived d[10]; d += 1;` for which I have a test.

File: clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.cpp
Comment (Discookie) [reply clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.cpp]: The issue may happen, but the check is only finding polymorphic objects, so no point in being more generic.

File: clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.cpp
Comment (Discookie) [reply clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.cpp]: I added the type alias matches as requested, but the pure virtual functions matcher has the same limitation as the regular virtual functions matcher - it cannot transitively decide if a function is pure or not, so it's quite limited as well. Added tests and a FIXME for it.

File: clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.h
Comment (Discookie) [reply clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.h]: Added.

File: clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.h
Comment (Discookie) [reply clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.h]: Fixed.

**Review by Discookie — COMMENTED** [review]
Renamed the check as requested, and fixed what I reasonably could regarding the rest.

Added the check option `MatchInheritedVirtualFunctions` to differentiate between inherited and newly-declared virtual functions. Inherited has many more false positives on the projects I tested.

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst
Comment (EugeneZelenko) [inline clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst:35]: Links to rules are usually placed at the end.

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst
Comment (EugeneZelenko) [inline clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst:42]: Please use single quotes for option values.

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst
Comment (EugeneZelenko) [inline clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst:45]: What is default value?

File: clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.cpp
Comment (5chmidti) [reply clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.cpp]: The `diag` member function only accepts a `SourceLocation` as the first parameter, that is correct.
When you stream a `SourceRange` into the diagnostic, that range will be underlined by the diagnostic engine (and in clangd). The same goes for attached fix-it hints, which highlight the ranges they change.

E.g.
```
  125 |   base += 1;
      |   ^
```
vs
```
  125 |   base += 1;
      |   ^~~~
```
when using 
```c++
  diag(PointerExpr->getBeginLoc(),
       "pointer arithmetic on polymorphic class '%0', which can result in "
       "undefined behavior if the pointee is a different class")
      << PointeeType->getName() 
      << PointerExpr->getSourceRange(); // adding this
```

File: clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.cpp
Comment (5chmidti) [reply clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.cpp]: Creating this matcher
```c++
AST_MATCHER(CXXRecordDecl, isAbstract) { return Node.isAbstract(); }
```
will provide a way to transitively know if a record is abstract, i.e. if any pure virtual function was not overridden, which is a guaranteed way to find out if it should be considered polymorphic. The matcher can simply be added inside an `anyOf` together with the `hasMethod` in both pointer expr matchers (`anyOf(isAbstract(), hasMethod(...))`). (I realize I didn't even mention `isAbstract` in the above comment in the inheritance part, which was the whole reason for that section)

Also, please move the `unless(isFinal())` matcher before the `hasMethod` matcher for slightly better performance.

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst
Comment (Discookie) [reply clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst:35]: Even after the `Options` subheading? `bugprone-reserved-identifier` has it before the `Options`, would look weirder at the end of it.

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst
Comment (Discookie) [reply clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst:45]: Default `false`, added.

File: clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.cpp
Comment (Discookie) [reply clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.cpp]: Tidy supports source ranges like that!? I never knew, thank you! Added.
I don't think I've seen support for it in CodeChecker, that's why I was surprised. Will have to follow up on that.

File: clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.cpp
Comment (Discookie) [reply clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.cpp]: Added `isAbstract`, and fixed `unless(isFinal())` as well (although with the amount of `final` classes in my test projects, it might even be slower in practice...).

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst
Comment (EugeneZelenko) [reply clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst:35]: Yes.

File: clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.cpp
Comment (5chmidti) [reply clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.cpp]: > with the amount of final classes in my test projects, it might even be slower in practice...

Hm, good point.

I ran this check with `time run-clang-tidy -clang-tidy-binary ./bin/clang-tidy -checks="-*,*polymorphic*" CodeGen` and couldn't measure any real difference when swapping around any of the matchers inside the `cxxRecordDecl` around. So maybe this isn't needed. I was originally about to suggest moving `isAbstract` before `hasMethod`, but that didn't matter either, at least in llvm. If you measure a noticeable difference in another project, then you can probably adjust the order of the matchers.

File: clang-tools-extra/clang-tidy/bugprone/PointerArithmeticOnPolymorphicObjectCheck.cpp
Comment (5chmidti) [inline clang-tools-extra/clang-tidy/bugprone/PointerArithmeticOnPolymorphicObjectCheck.cpp:30]: Please store the value of `MatchInheritedVirtualFunctions`, instead of `true`.

**Review by 5chmidti — APPROVED** [review]
LGTM minus the stored option value. (please wait for others to finish their review)

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst:40]: this name and description is confusing.
Maybe name it like: IgnoreInheritedVirtualFunctions
Check should be default match all polymorphic classes, and then options could be just to reduce scope, like ignore classes with only virtual destructor, and so on.

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst:42]: Maybe: "When `true`, all polymorphic classes are checked regardless if they add new virtual functions or not. Default value is `false`"

File: clang-tools-extra/clang-tidy/bugprone/PointerArithmeticOnPolymorphicObjectCheck.h
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/bugprone/PointerArithmeticOnPolymorphicObjectCheck.h:31]: Exclude implicit code with TK_IgnoreUnlessSpelledInSource here (look into other checks).

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (PiotrZSL) [inline clang-tools-extra/docs/ReleaseNotes.rst:132]: Change this description to point about polymorphic-object. Or instead of that declare (class does not declare anything), just say "that contain".

File: clang-tools-extra/clang-tidy/bugprone/PointerArithmeticOnPolymorphicObjectCheck.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/bugprone/PointerArithmeticOnPolymorphicObjectCheck.cpp:39]: 1. you dont need to repeat cxxRecordDecl.
2. Create custom AST matcher that utilize [isPolymorphic](https://clang.llvm.org/doxygen/classclang_1_1CXXRecordDecl.html#a6989371936079ccd01556c99637e9de0) ()

File: clang-tools-extra/clang-tidy/bugprone/PointerArithmeticOnPolymorphicObjectCheck.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/bugprone/PointerArithmeticOnPolymorphicObjectCheck.cpp:56]: what about /=, *=, /, *

File: clang-tools-extra/clang-tidy/bugprone/PointerArithmeticOnPolymorphicObjectCheck.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/bugprone/PointerArithmeticOnPolymorphicObjectCheck.cpp:63]: that's too generic, register each of those matchers separately.

File: clang-tools-extra/clang-tidy/bugprone/PointerArithmeticOnPolymorphicObjectCheck.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/bugprone/PointerArithmeticOnPolymorphicObjectCheck.cpp:70]: This may fail with typedefs, you already got cxxRecordDecl in matcher, just bind it to some name, like "Record"

File: clang-tools-extra/clang-tidy/bugprone/PointerArithmeticOnPolymorphicObjectCheck.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/bugprone/PointerArithmeticOnPolymorphicObjectCheck.cpp:73]: "... on polymorphic object of type '%0'"

File: clang-tools-extra/clang-tidy/bugprone/PointerArithmeticOnPolymorphicObjectCheck.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/bugprone/PointerArithmeticOnPolymorphicObjectCheck.cpp:47]: This doesn't make sense, simply IF type is derived with added new methods, then there will be no issues, as vtable will handle everything correctly, and even if operation would be done on base class everything will be fine.

Actually only issue will be if size of classes does not match, but that can happen whatever this is polymorphic object of not.

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst:10]: "as the two types can have different sizes."
this apply also to non polymorphic classes.

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst:35]: wrap this line,. move link to new line.

File: clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.cpp
Comment (PiotrZSL) [reply clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.cpp]: This should be default scenario, to detect all usage on polymorphic classes, and options should reduce scope.
I personally do not see use case for pointer arithmetic on such classes.

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/virtual-arithmetic.rst
Comment (PiotrZSL) [reply clang-tools-extra/docs/clang-tidy/checks/bugprone/virtual-arithmetic.rst]: make sure just to write that it may cause crash as due to size mismatch vtable pointer may point to invalid  address.

File: clang-tools-extra/test/clang-tidy/checkers/bugprone/virtual-arithmetic.cpp
Comment (PiotrZSL) [reply clang-tools-extra/test/clang-tidy/checkers/bugprone/virtual-arithmetic.cpp]: Object* ptr = &table[5];

**Review by PiotrZSL — CHANGES_REQUESTED** [review]
Few nits:
- storeOptions needs fix
- Consider excluding implicit code with TK_IgnoreUnlessSpelledInSource (not this may exclude also template instances, so I leave it up to you)
- Documentation require small work
- Make check more strict by default, and add options to relax it (as thats what [CTR56-CPP](https://wiki.sei.cmu.edu/confluence/display/cplusplus/CTR56-CPP.+Do+not+use+pointer+arithmetic+on+polymorphic+objects) require). I undestand that on your project this may cause some false-positives, but still thats your project problem. Consider adding options to "Allow classes where only virtual function is destructor", "Classes where all virtual functions are marked final",  "Classes that do not add new virtual functions or members". (Just check what suit your project most).

Note that next release branch out is in a ~month. Would be nice if this check could fit-in.

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (PiotrZSL) [inline clang-tools-extra/docs/ReleaseNotes.rst:129]: add in line 173 info about new alias (check release notes from previous release)

File: clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.cpp
Comment (Discookie) [reply clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.cpp]: I thought about making it default for the CERT alias only, but in the end made it default for all aliases. Didn't really find a compelling false-positive that wasn't project specific.

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/virtual-arithmetic.rst
Comment (Discookie) [reply clang-tools-extra/docs/clang-tidy/checks/bugprone/virtual-arithmetic.rst]: Added.

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst
Comment (Discookie) [reply clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst]: Changed the description along the name.

**Review by Discookie — COMMENTED** [review]
I found a curious report when browsing through the results for `llvm-project`: [MemoryBuffer.cpp](https://codechecker-demo.eastus.cloudapp.azure.com/Default/report-detail?review-status=Unreviewed&review-status=Confirmed%20bug&detection-status=New&detection-status=Reopened&detection-status=Unresolved&run=llvm-project_llvmorg-12.0.0_ctr56-allow-inherited&report-id=5498248&report-hash=2e0e0ee19429584af2a682c9efb1cdf4&report-filepath=%2Flocal%2Fpersistent_docker%2FCSA-measurements-driver-2931%2Fmeasurements_workspace%2Fllvm-project%2Fllvm%2Flib%2FSupport%2FMemoryBuffer.cpp)
I have no idea where these spurious fixits could be coming from. My check doesn't emit any. Could this be because of the source range I'm piping into my diagnostic, or is it just some side-effect of other checks, or maybe the Tidy module? I'll investigate and fix this before merging.

Other than this case, fixed the reviews. I'm debating adding a flag for virtual destructor only as well, but I haven't seen a case in the wild where it was the only inherited function.

Made IgnoreInheritedVirtualFunctions `false` by default - doesn't have too many meaningful differences, but it does have more reports. [Results on projects](https://codechecker-demo.eastus.cloudapp.azure.com/Default/runs?run=ctr56%2ainherited)

**Comment by Discookie** [comment]
Seems to be a CodeChecker-specific issue, doesn't block upstreaming this review.

File: clang-tools-extra/clang-tidy/bugprone/PointerArithmeticOnPolymorphicObjectCheck.cpp
Comment (5chmidti) [inline clang-tools-extra/clang-tidy/bugprone/PointerArithmeticOnPolymorphicObjectCheck.cpp:77]: Nit/Idea: 
- remove `, which`
- `if the pointee is a different object` -> `if the dynamic type is different` or `if the dynamic type differs from the pointer type`

Comment (Discookie) [reply clang-tools-extra/clang-tidy/bugprone/PointerArithmeticOnPolymorphicObjectCheck.cpp]: Seems much clearer, fixed.

**Review by PiotrZSL — APPROVED** [review]
Some small issues with diagnostic left.
Except that, looks good enough.

File: clang-tools-extra/clang-tidy/bugprone/PointerArithmeticOnPolymorphicObjectCheck.cpp
Comment (Discookie) [reply clang-tools-extra/clang-tidy/bugprone/PointerArithmeticOnPolymorphicObjectCheck.cpp]: Whoops that's a typo, well spotted! Fixed.

**Pull request merged**
